design/one-pager-configuration-stacks.md

```diff
@@ -0,0 +1,282 @@
+* A base boilerplate tooling where creating a new configuration set that has a
+  controller is as easy as changing the YAML files.
+* An easy way for users to deploy a set of pre-defined configurations that has
+  a controller which reconciles the set of the resources deployed.
+
+## Proposal
+
+Crossplane has the notion of stacks that can be used as extension to Crossplane's
+abilites, to deploy an application and all other things that you can do with
+Kubernetes controllers since main components of stacks are CRDs and controllers
+watching them.
+
+To achieve the mentioned goals, we can publish the YAML files in a stack with a
+controller that has its own CRD with high level parameters. Controller deploys
+and updates the YAMLs that are packaged in the stack image according to user's
+input on the CR instance of that stack that user creates.
+
+As sky is the limit when it comes to stacks, we need to put down some general
+rules to be adhered by initial configuration stacks.
+
+* It should be possible to have different sets of configurations in one stack
+  and user should be able to choose from them via the CR that is an instance
+  of stack's CRD.
+  * Note that it's highly preferred to have only one custom resource definition that
+    the stack uses. You can expose the high level variations of the parameters
+    through `spec` instead of having different custom resource definitions in the
+    same stack or you can write another stack with its own CRD for that purpose.
+* A minimal high level set of configurations are exposed to the user through
+  the stack's CR, such as reference to the cloud credential secret or `region`.
+* All resources should be labelled referring to the CR instance of the stack.
+* In YAMLs, resources should refer to each other using cross-resource references
+  wherever possible.
+* Controller should update the resources continuously and treat the given CR as
+  the source of truth even though user manually changes the resources that are
+  deployed by the stack.
+* CRD of the configuration stack should be cluster-scoped since the author of
+  the CR instances of the configuration stack is assumed to be the infrastructure
+  owner.
+* Stack type will be `ClusterStack` as most of the resources it deploys are
+  cluster-scoped.
+
+## User Experience
+
+1. Install AWS stack and create a secret that contains AWS credentials.
+2. Install a AWS configuration stack via:
+```yaml
+# exact apiVersion is TBD.
+apiVersion: stacks.crossplane.io/v1alpha1
+kind: ClusterStackInstall
+metadata:
+  name: minimal-aws
+  namespace: crossplane-system
+spec:
+  package: "crossplane/minimal-aws:latest"
+```
+
+Now, I want an environment where all my database instances and kubernetes clusters
+are connected to each other in a private VPC, which what this specific configuration
+stack does. Create the following:
+
+```yaml
+apiVersion: aws.configurationstacks.crossplane.io/v1alpha1
+kind: MinimalAWS
+metadata:
+  name: small-infra
+spec:
```

Comment:
We'll want to support creating resource classes marked as default, but if more than one MinimalAWS CRs are created, there should be a way to disable the default label being applied so there is only 1 set of default classes.

Reply:
I have added a section about this. I suggest we have a field named `keepDefaultingLabels`(name suggestions are welcome) and if user intends to keep the defaulting label in resource classes, they have to make this field `true`. Otherwise, the controller will remove the defaulting labels whenever it sees one. So, by default, if several users deploy empty CRs they won't have any conflict. If they want the default, they'd need to explicitly say so.